homeassistant/components/mqtt/cover.py

```diff
@@ -333,11 +333,12 @@ def position_message_received(msg):
                     float(payload), COVER_PAYLOAD
                 )
                 self._position = percentage_payload
-                self._state = (
-                    STATE_CLOSED
-                    if percentage_payload == DEFAULT_POSITION_CLOSED
-                    else STATE_OPEN
-                )
+                if self._config.get(CONF_STATE_TOPIC) is None:
```

Comment:
```suggestion
                if self._config.get(CONF_STATE_TOPIC) is None or self._position is None:
```

Reply:
@CubieMedia can you explain the logic for this `or self._position is None`, we only reach here if we have a numeric payload.

Reply:
This solves a problem that exists on restarting Home Assistant. The device sends the stop state (cover is not moving) but we can not determine the real state (closed, open), for this we need the position. So what the change does, it takes the first position message (self._position is None) to also set the state.

> we only reach here if we have a numeric payload.
This is correct and should stay that way, the code only differs if self._position has been set before. That is why we have to set position afterwards. Otherwise we would not know if it has been set.